Add starting point for ByteArray-backed bytecode generator

austnwil opened the pull request:
*Description of changes:*

This starts the implementation of the binary 1.1 array-backed bytecode generator, adding a basic refill implementation and support for reading integer, lob, text, and short timestamp references.

Reading decimal and long timestamp references is not yet implemented.

I pulled the test cases for the opcode handlers into shared test cases since they were also useful for testing the generator. A lot of files were changed but most of them are just refactoring the existing opcode handler tests to use the shared test cases.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

austnwil committed:
Begin implementation of ByteArrayBytecodeGenerator
- Implement readTextReference
- Implement readBytesReference
- Implement basic refill

austnwil committed:
Implement readBigIntegerReference

austnwil committed:
Implement readShortTimestampReference

austnwil committed:
Rename UTF8 decoder

austnwil committed:
Add tests; fix bug in readBytesReference

austnwil committed:
Add comments

austnwil committed:
Add test for bytecode generator; refactor opcode handler test cases
- Opcode handler tests now use the common test cases shared by the
  bytecode generator tests
- Added test to bytecode generator that tests long inputs containing all
  supported opcodes

austnwil committed:
Fix broken refill logic; adjust comments

austnwil committed:
Minor fixes

austnwil committed:
Remove unnecessary test cases causing OOM errors

popematt commented on src/main/java/com/amazon/ion/bytecode/BytecodeEmitter.kt:
```diff
@@ -87,4 +87,14 @@ internal object BytecodeEmitter {
+    @JvmStatic
+    fun emitEndOfInput(destination: BytecodeBuffer) {
+        destination.add(Instructions.I_END_OF_INPUT)
+    }
```
Not a blocker.

We don't need to have helper functions for everything. In this case, it's worth asking what is this actually helping with? Some of the other helper functions in the class handle things like spreading the bits for a long value or converting from float to the int bits.

If you compare these two, I think the version with `BytecodeHelper` doesn't add value, and maybe even obscures what's going on.
```kotlin
destination.add(Instructions.I_REFILL)

BytecodeHelper.emitRefill(destination)
```

austnwil replied:
Good point

popematt commented on src/main/java/com/amazon/ion/bytecode/bin11/ByteArrayBytecodeGenerator11.kt:
```diff
@@ -2,14 +2,99 @@
+        }
+
+        if (currentPosition < source.size) {
+            BytecodeEmitter.emitRefill(destination)
```
I attempted to work on writing tests for a different `BytecodeGenerator` implementation, and I found that it added a lot of noise to have the `REFILL` instruction added by the `BytecodeGenerator`. It's also something that's the same regardless of the implementation, and it's a concern of the reader rather than being part of the data, so I think it makes more sense for the `BytecodeIonReader` to handle adding in the `REFILL` instruction.

popematt commented on src/main/java/com/amazon/ion/bytecode/bin11/bytearray/ShortTimestampDecoder.kt:
```diff
@@ -0,0 +1,221 @@
+/**
+ * Helper class for decoding the various short timestamp encoding variants from a [ByteArray].
+ */
+internal object ShortTimestampDecoder {
```
How about just `TimestampDecoder`? There's not really a need to have separate short and long timestamp decoder classes.

popematt commented on src/main/java/com/amazon/ion/bytecode/bin11/bytearray/ShortTimestampDecoder.kt:
```diff
@@ -0,0 +1,221 @@
+        return Timestamp.forSecond(year + 1970, month, day, hour, minute, secondBigDecimal.add(fractionalSecondBigDecimal), (offset - 56) * 15)
+    }
+
+    fun readTimestamp(source: ByteArray, position: Int, precisionAndOffsetMode: Int): Timestamp {
```
Is `precisionAndOffsetMode` the `opcode`?

austnwil replied:
It was just the low nibble. Looks like the bytecode reference does call for the entire opcode, so I made that change

popematt commented on src/main/java/com/amazon/ion/bytecode/bin11/bytearray/ShortTimestampDecoder.kt:
```diff
@@ -0,0 +1,221 @@
+    fun readTimestamp(source: ByteArray, position: Int, precisionAndOffsetMode: Int): Timestamp {
+        // TODO: calling function references like this might be slower than just using a conditional or other solutions.
+        //  Might be worth looking into.
+        val decoder = opcodeToDecoderFunctionTable[precisionAndOffsetMode]
```
Yes, I think it will be. A `when` on the opcode is probably going to be faster because it will be compiled as a `tableswitch` instruction in the java bytecode instead of having to load the array, and then load a value from the array (both of which require fetching objects from the heap), and _then_ the JVM potentially still has to perform a vtable lookup to find the correct method implementation for the interface.

The extra indirection here also limits the JVM's ability to perform method inlining. For example, if an application only reads data that uses 0x85 opcode, the JVM could inline that particular method into this method, eliminating the overhead of the function call. However, when you're using method references like this, hinders that ability.

FYI, I think I wrote a note about that here: https://github.com/popematt/ion-java/blob/6f5274ecc5e5812f08884a6d1aa1c4d7546861fe/src/main/java/com/amazon/ion/v8/TimestampHelper.kt#L34-L38

So why are use using an array of function references for the opcode lookup? The main reasons are that (1) the opcode handler table is already going to be hard for the JVM to predict anyway because there's going to be lots of different opcodes in normal Ion data, (2) the `when` would make the method very long and hard to read, and (3) the compiled method is going to be quite large and unable to be inlined at its own call sites.

austnwil replied:
Great explanation!

popematt commented on src/test/java/com/amazon/ion/TextToBinaryUtils.kt:
```diff
@@ -64,4 +81,29 @@ object TextToBinaryUtils {
+    @JvmStatic
+    fun String.decimalStringToIntArray(): IntArray {
+        return octetStringToIntArray(this, 10)
+    }
```
Why is this necessary when you can just do `intArrayOf(105, -9349549, 0, -12, 99999)`?

popematt commented on src/test/java/com/amazon/ion/TextToBinaryUtils.kt:
```diff
@@ -64,4 +81,29 @@ object TextToBinaryUtils {
+    @JvmStatic
+    fun ByteArray.byteArrayToHexString(): String {
+        return this.joinToString(" ") { it.toUByte().toString(16).padStart(2, '0') }
+    }
```
Just use `ByteArray.toHexString()`, provided in the Kotlin standard library, instead of adding this method.

popematt commented on src/test/java/com/amazon/ion/bytecode/bin11/OpcodeTestCases.kt:
```diff
@@ -0,0 +1,453 @@
+ * sensitive to the opcode's position in the input (e.g. `OP_*_REF` codes) to be reused across test cases that use them
+ * at different offsets. Pass the decimal string to [replacePositionTemplates] to parse these placeholders.
+ */
+object OpcodeTestCases {
```
Can we wait until these actually need to be re-used by multiple bytecode generators before we do something like this. Yes, I know I created test cases for FlexInts and FixedInts, but that was already being used for both the encoding and decoding side, and they were being used for testing multiple methods that only varied in whether they accepted/returned an `int`, `long`, or `BigInteger`.

I'm not convinced of the re-usability here because there isn't a general requirement that specific opcodes must correspond to a specific bytecode instruction. It would be valid for an inline symbol to be compiled as a `I_SYMBOL_REF`, `I_SYMBOL_CP`, or even a `I_SYMBOL_CHAR` (if it's only one character long).

Furthermore, the string substitution mechanism is adding complexity, which makes me inclined to think that we should wait until we discover that we are duplicating our tests cases a lot.

austnwil replied:
> there isn't a general requirement that specific opcodes must correspond to a specific bytecode instruction...

This is a great point, hadn't thought of that.

popematt commented on src/test/java/com/amazon/ion/bytecode/bin11/ByteArrayBytecodeGenerator11Test.kt:
```diff
@@ -0,0 +1,99 @@
+import com.amazon.ion.bytecode.bin11.OpcodeTestCases.STRING_REFERENCE_OPCODE_CASES
+import com.amazon.ion.bytecode.bin11.OpcodeTestCases.TYPED_NULL_OPCODE_CASES
+import com.amazon.ion.bytecode.bin11.OpcodeTestCases.replacePositionTemplates
+import com.amazon.ion.bytecode.ir.Instructions
+import com.amazon.ion.bytecode.util.BytecodeBuffer
+import com.amazon.ion.bytecode.util.ConstantPool
+import org.junit.jupiter.api.Assertions.assertArrayEquals
+import org.junit.jupiter.api.Assertions.assertEquals
+import org.junit.jupiter.params.ParameterizedTest
+import org.junit.jupiter.params.provider.MethodSource
+
+class ByteArrayBytecodeGenerator11Test {
+
+    @ParameterizedTest
+    @MethodSource(
+        BOOLEAN_OPCODE_CASES, NULL_OPCODE_CASES, TYPED_NULL_OPCODE_CASES, FLOAT0_OPCODE_CASES,
+        FLOAT16_OPCODE_CASES, FLOAT32_OPCODE_CASES, FLOAT64_OPCODE_CASES, SHORT_TIMESTAMP_OPCODE_CASES,
+        REFERENCE_OPCODE_CASES, INT0_OPCODE_CASES, INT8_OPCODE_CASES, INT16_OPCODE_CASES, INT24_OPCODE_CASES,
+        INT32_OPCODE_CASES, INT64_EMITTING_OPCODE_CASES, STRING_REFERENCE_OPCODE_CASES, LOB_REFERENCE_OPCODE_CASES
+    )
```
I don't think that it's necessary to repeat all of the tests for each individual opcode. The tests at this level should focus on some cases that are composed of multiple opcodes, cases where there are directives, macro compilation, etc.

popematt commented on src/test/java/com/amazon/ion/bytecode/bin11/ByteArrayBytecodeGenerator11Test.kt:
```diff
@@ -0,0 +1,99 @@
+        val timestampPosition = bytecode.get(1)
+        val expectedTimestamp = Timestamp.valueOf(expectedTimestampString)
+        val readTimestamp = generator.readShortTimestampReference(timestampPosition, timestampPrecisionAndOffsetMode)
+        assertEquals(expectedTimestamp, readTimestamp)
```
I realize I said previously that we should use them how they're meant to be used, but I guess I didn't qualify that very well. Let's test the `read*Reference()` methods in isolation from the `refill()` method so that it's easier to track down a failure if something goes wrong.

E.g.:
```kotlin
val bytes = timestampReferenceBytes.hexStringToByteArray()
val generator = ByteArrayBytecodeGenerator11(bytes, 0)
val opcode = bytes[0].toInt().and(0xFF)
val readTimestamp = generator.readShortTimestampReference(1, opcode)
assertEquals(expectedTimestamp, readTimestamp)
```

popematt commented on src/test/java/com/amazon/ion/bytecode/bin11/bytearray/PrimitiveDecoderTest.kt:
```diff
@@ -160,6 +161,18 @@ class PrimitiveDecoderTest {
+    @ParameterizedTest
+    @MethodSource(FIXED_INT_8_CASES, FIXED_INT_16_CASES, FIXED_INT_24_CASES, FIXED_INT_32_CASES, FIXED_INT_64_CASES)
+    @CsvSource(
+        " 9223372036854775808, 00000000 00000000 00000000 00000000 00000000 00000000 00000000 10000000 00000000",
+        "-9223372036854775809, 11111111 11111111 11111111 11111111 11111111 11111111 11111111 01111111 11111111",
```
We can add in some over-padded values too.
```suggestion
        " 9223372036854775808, 00000000 00000000 00000000 00000000 00000000 00000000 00000000 10000000 00000000",
        "-9223372036854775809, 11111111 11111111 11111111 11111111 11111111 11111111 11111111 01111111 11111111",
        "                   1, 00000001 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000",
        "                  -1, 11111111 11111111 11111111 11111111 11111111 11111111 11111111 11111111 11111111",
```

popematt commented on src/main/java/com/amazon/ion/bytecode/bin11/bytearray/ShortTimestampDecoder.kt:
```diff
@@ -0,0 +1,221 @@
+import java.math.BigDecimal
+
+/**
+ * Helper class for decoding the various short timestamp encoding variants from a [ByteArray].
```
Small TODO note for later.

```suggestion
 * Helper class for decoding the various short timestamp encoding variants from a [ByteArray].
 *
 * TODO(perf): avoid auto-boxing the `0` integer for the offset when constructing the Timestamp instance.
```

popematt commented on src/main/java/com/amazon/ion/bytecode/bin11/bytearray/ShortTimestampDecoder.kt:
```diff
@@ -0,0 +1,221 @@
+    fun readTimestampToMonth(source: ByteArray, position: Int): Timestamp {
+        val yearAndMonth = readFixedInt16(source, position).toInt()
+        val year = yearAndMonth.and(MASK_7)
+        val month = yearAndMonth.shr(7)
```
FYI, `com.amazon.ion.impl.bin.Ion_1_1_Constants` has constants for all of these offsets so that we don't need to be using "magic" numbers here.

popematt commented on src/test/java/com/amazon/ion/bytecode/bin11/bytearray/ReferenceOpcodeHandlerTest.kt:
```diff
@@ -2,145 +2,38 @@
+            OpCode.VARIABLE_LENGTH_SYMBOL -> ReferenceOpcodeHandler(Instructions.I_SYMBOL_REF)
+            OpCode.VARIABLE_LENGTH_BLOB -> ReferenceOpcodeHandler(Instructions.I_BLOB_REF)
+            OpCode.VARIABLE_LENGTH_CLOB -> ReferenceOpcodeHandler(Instructions.I_CLOB_REF)
+            else -> fail("Opcode is not a variable-length reference opcode: 0x${opcode.toSingleHexByte()}")
```
Or you could just use the standard library functions.
```suggestion
            else -> fail("Opcode is not a variable-length reference opcode: 0x${opcode.toByte().toHexString()}")
```

austnwil committed:
Apply suggestions from code review

Co-authored-by: Matthew Pope <81593196+popematt@users.noreply.github.com>

austnwil committed:
Implement more PR suggestions

austnwil committed:
PR suggestion: Revert refactoring of opcode handler test cases

austnwil committed:
Don't emit I_REFILL in generator

austnwil committed:
Add generator tests

austnwil commented:
Thanks for the thorough review! I added some changes; I believe I got everything you mentioned. Here are the new updates:

- Removed the superfluous `emitRefill` and `emitEndOfInput` helpers
- Stopped emitting `I_REFILL` in the generator
- Renamed `ShortTimestampDecoder` to `TimestampDecoder`
- Changed the short timestamp-related logic to store and read the entire opcode in the bytecode rather than just the low nibble
- Removed the function lookup table in the short timestamp decoder in favor of a `when`
- Reverted the refactoring of the opcode handler tests that brought the test cases into `OpcodeTestCases`
- Created a higher-level test for the bytecode generator instead of just retesting the handlers
- Removed some unnecessary helper funcs
- Applied some other small suggestions and fixes

popematt commented on src/main/java/com/amazon/ion/bytecode/bin11/ByteArrayBytecodeGenerator11.kt:
```diff
@@ -2,14 +2,97 @@
+        macroIndices: IntArray,
+        symTab: Array<String?>
+    ) {
+        var opcode = 0
```
0 is a valid opcode. It's probably a little clearer/safer to initialize it to something that is not a valid opcode.
```suggestion
        var opcode = -1
```

austnwil replied:
Good idea

popematt commented on src/main/java/com/amazon/ion/bytecode/bin11/ByteArrayBytecodeGenerator11.kt:
```diff
@@ -2,14 +2,97 @@
+
+    override fun ionMinorVersion(): Int {
+        return 1
+    }
```
You don't have to do this, but I thought I'd just mention it to make sure you were aware that Kotlin lets you write single-expression functions like this.
```suggestion
    override fun ionMinorVersion(): Int = 1
```

popematt commented on src/test/java/com/amazon/ion/bytecode/bin11/ByteArrayBytecodeGenerator11Test.kt:
```diff
@@ -0,0 +1,138 @@
+                "6F " // false
+        ]
+    )
+    fun `generator can compile input containing multiple simple opcodes`(inputBytesString: String) {
```
A parameterized test looks a little fishy when there's only one case provided.
Also, we have a utility function that allows you to have comments in hex strings so that we can use kotlin multiline strings instead of string concatenation. It just makes it a little easier to read.
```suggestion
    @Test
    fun `generator can compile input containing multiple simple opcodes`() {
        val inputBytesString = """
            64 4F 97 21 C5             | int -987654321
            86 35 7D CB 12 2E 22 1B    | timestamp 2023-10-15T11:22:33.444555-00:00
            8F 0C                      | null.struct
            6A                         | float 0e0
            6D 18 2D 44 54 FB 21 09 40 | float 3.141592653589793
            FE 31                      | 24-byte blob
            49 20 61 70 70 6c 61 75    |
            64 20 79 6f 75 72 20 63    |
            75 72 69 6f 73 69 74 79    |
            6F                         | false
        """.cleanCommentedHexBytes()
```

austnwil replied:
Forgot about that helper, good call

popematt commented on src/test/java/com/amazon/ion/bytecode/bin11/bytearray/TimestampDecoderTest.kt:
```diff
@@ -0,0 +1,39 @@
+        "8A 35 7D CB EA 85 BC 01,        2023-10-15T11:22:33.444+01:15",
+        "8B 35 7D CB EA 85 8B C8 06,     2023-10-15T11:22:33.444555+01:15",
+        "8C 35 7D CB EA 85 92 61 7F 1A,  2023-10-15T11:22:33.444555666+01:15",
+        // TODO: add min/max values, other extremes
```
FYI, we have a bunch of test cases here that you could add—just need to convert from binary to hex: https://github.com/popematt/ion-java/blob/6f5274ecc5e5812f08884a6d1aa1c4d7546861fe/src/test/java/com/amazon/ion/impl/bin/IonEncoder_1_1Test.java#L337-L390

austnwil replied:
Added more test cases and they actually caught a bug

austnwil committed:
Replace UTF-8 test string with known source

austnwil committed:
Apply suggestion from @popematt

Co-authored-by: Matthew Pope <81593196+popematt@users.noreply.github.com>

austnwil committed:
Apply suggestion from @popematt

Co-authored-by: Matthew Pope <81593196+popematt@users.noreply.github.com>

austnwil committed:
Apply suggestion from @popematt

Co-authored-by: Matthew Pope <81593196+popematt@users.noreply.github.com>

austnwil committed:
Fix broken bytecode generator test

austnwil committed:
Add various test cases to timestamp decoder

austnwil committed:
Fix bug in day-precision timestamp decoder when day has high bit set

austnwil committed:
Merge branch 'ion11' into austnwil/ion11-bytearray-bytecode-generator
